backend/danswer/background/connector_deletion.py

```diff
@@ -0,0 +1,296 @@
+from danswer.db.document import delete_document_by_connector_credential_pair
+from danswer.db.document import delete_documents_complete
+from danswer.db.document import get_chunk_ids_for_document_ids
+from danswer.db.document import (
+    get_chunks_with_single_connector_credential_pair,
+)
+from danswer.db.document import (
+    get_document_by_connector_credential_pairs_indexed_by_multiple,
+)
+from danswer.db.engine import get_sqlalchemy_engine
+from danswer.db.index_attempt import delete_index_attempts
+from danswer.db.models import Credential
+from danswer.db.models import DeletionAttempt
+from danswer.db.models import DeletionStatus
+from danswer.utils.logger import setup_logger
+
+logger = setup_logger()
+
+
+def _delete_connector_credential_pair(
+    db_session: Session,
+    vector_index: VectorIndex,
+    keyword_index: KeywordIndex,
+    deletion_attempt: DeletionAttempt,
+) -> int:
+    connector_id = deletion_attempt.connector_id
+    credential_id = deletion_attempt.credential_id
+
+    def _delete_singly_indexed_docs() -> int:
+        # if a document store entry is only indexed by this connector_credential_pair, delete it
+        num_docs_deleted = 0
+        chunks_to_delete = get_chunks_with_single_connector_credential_pair(
+            db_session=db_session,
+            connector_id=connector_id,
+            credential_id=credential_id,
+        )
+        if chunks_to_delete:
+            document_ids: set[str] = set()
+            vector_chunk_ids_to_delete: list[str] = []
+            keyword_chunk_ids_to_delete: list[str] = []
+            for chunk in chunks_to_delete:
+                document_ids.add(chunk.document_id)
+                if chunk.document_store_type == StoreType.KEYWORD:
+                    keyword_chunk_ids_to_delete.append(chunk.id)
+                else:
+                    vector_chunk_ids_to_delete.append(chunk.id)
+
+            vector_index.delete(ids=vector_chunk_ids_to_delete)
+            keyword_index.delete(ids=keyword_chunk_ids_to_delete)
+            # removes all `Chunk`, `DocumentByConnectorCredentialPair`, and `Document`
+            # rows from the DB
+            delete_documents_complete(
+                db_session=db_session,
+                document_ids=list(document_ids),
+            )
+            num_docs_deleted += len(document_ids)
+        return num_docs_deleted
+
+    num_docs_deleted = _delete_singly_indexed_docs()
+
+    def _update_multi_indexed_docs() -> None:
+        # if a document is indexed by multiple connector_credential_pairs, we should
+        # update it's access rather than outright delete it
+        document_by_connector_credential_pairs_to_update = (
+            get_document_by_connector_credential_pairs_indexed_by_multiple(
+                db_session=db_session,
+                connector_id=connector_id,
+                credential_id=credential_id,
+            )
+        )
+
+        def _get_user(
```

Comment:
slight preference to have these helper functions declared at the top of the function rather than interwoven into it